### PR TITLE
Custom op interface to the C API to remove shared library dependency

### PR DESCRIPTION
**Pull request by RyanUnderhill**
It's not complete yet but this is to get feedback before going further

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -541,7 +557,7 @@ struct OrtCustomOp {
   uint32_t version;  // Initialize to ORT_API_VERSION
 
   // This callback creates the kernel, which is a user defined parameter that is passed to the Kernel* callbacks below.
-  void(ORT_API_CALL* CreateKernel)(_In_ struct OrtCustomOp* op, _In_ OrtKernelInfo* info, _Out_ void** op_kernel);
+  void(ORT_API_CALL* CreateKernel)(_In_ struct OrtCustomOp* op, _In_ const OrtCustomOpApi* api, _In_ OrtKernelInfo* info, _Out_ void** op_kernel);
```
Comment (snnn) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: _In_ const OrtKernelInfo* info
or remove the _In_.

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 2]
```diff
@@ -177,34 +177,37 @@ template <typename T, size_t N>
...
+    OrtTensorDimensions dimensions(api_, inputs[0]);
+    ORT_THROW_ON_ERROR(api_.OrtSetDims(info, dimensions.data(), dimensions.size()));
   }
 
   void Compute(OrtValue** inputs, size_t /*input_count*/, OrtValue** outputs, size_t /*output_count*/) {
     const float* X;
     const float* Y;
-    ORT_THROW_ON_ERROR(OrtGetTensorMutableData(inputs[0], reinterpret_cast<void**>(const_cast<float**>(&X))));
-    ORT_THROW_ON_ERROR(OrtGetTensorMutableData(inputs[1], reinterpret_cast<void**>(const_cast<float**>(&Y))));
+    ORT_THROW_ON_ERROR(api_.OrtGetTensorMutableData(inputs[0], reinterpret_cast<void**>(const_cast<float**>(&X))));
```
Comment (snnn) [inline onnxruntime/test/shared_lib/test_inference.cc]: Well, as we talked before, I **hate** this. If I were the developer, if you ask me, do I want to put something in DLLMain, or rewrite the whole code base to change 'OrtXXX' to 'api.OrtXXX', I would definitely choose the first one.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: I made it const and fixed up the downstream effects of it too.

File: onnxruntime/test/shared_lib/test_inference.cc (hunk 2 above)
Comment (RyanUnderhill) [reply onnxruntime/test/shared_lib/test_inference.cc]: I made a small renaming change to help a bit.

**Pull request merged**